### PR TITLE
[WIP] Changed typescript definitions to cover .jsx files

### PR DESCRIPTION
**Pull request by cromefire**
Because not everyone wants pretranspiled files

### CODE REVIEW

**Comment by prateekbh** [comment]
request @YePpHa to review this... cuz he's the TS expert here

**Comment by cromefire** [comment]
Changes I made:

- `index.d.ts` -> `Component.d.ts`
- new `index.d.ts` which exports everything from `Component.d.ts`

Just allows importing `Component.jsx` directly from typescript

**Comment by cromefire** [comment]
Because if you have your own infrastructure to handle jsx and es6+ it's not good to import the index.js files (I for example have builds for browsers with and without esmodule support)

**Comment by prateekbh** [comment]
makes sense, @YePpHa it'll be great if u can have a look too

**Comment by cromefire** [comment]
Okay after trying to build my app I found out that this isn't just nicer to let my own infrastructure handle this, but needed because I can't build with babel because of some errors with webpack and babel-runtime and not without babel because then I'm having errors with the classes not being callable without new (somehow babel requires this for inheritance).
So I'm blocked until this is merged, so please have a look @YePpHa it shouldn't be that complicated.

**Comment by artipost** [comment]
Failed pictures (1): [failed-pictures.zip](https://gist.githubusercontent.com/artipost/a9ab69bacf42da7c3c885e22d571f8c4/raw/64a619a8a919e79d13416d878583c8fc5c559b5b/failed-pictures.zip)

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [407548600](https://travis-ci.org/prateekbh/preact-material-components/builds/407548600), for commit [9618403](https://github.com/prateekbh/preact-material-components/pull/874/commits/9618403440af90d19228daa8f0de7ebcacb99d35).

**Comment by cromefire** [comment]
Seems like a faulty build

**Comment by cromefire** [comment]
It says the original image has a height of 0

**Comment by prateekbh** [comment]
I'll get this released by EOD

**Comment by cromefire** [comment]
Okay, there is a problem: When I import TextField.jsx it tries to import ../Icon(/index.js).
Is there any way to rewrite ../Icon/Icon to ../Icon at build time?

**Comment by prateekbh** [comment]
Its clearly a bug, it should import `/Icon/Icon.jsx`. Can you add the change to this PR?

**Comment by cromefire** [comment]
I started changing that right in this PR, but I can move this into a new one, I made a custom transform for it

**Comment by artipost** [comment]
Failed pictures (1): [failed-pictures.zip]

##
:postbox: Posted by [artipost.io](https://artipost.io), through Travis build [407671625](https://travis-ci.org/prateekbh/preact-material-components/builds/407671625), for commit [7269c83](https://github.com/prateekbh/preact-material-components/pull/874/commits/7269c837e889a641efcfd55aa0f12f9e00eee56d).

**Comment by prateekbh** [comment]
@cromefire can you work without this change for now?
I am planning to move to rollup every component instead of just babel transform, that way we can produce index.js -> es5 and index.esm.js -> ESM output for every component without a custom transformer.

Also would this affect the way we declare TS files?

**Comment by cromefire** [comment]
a) no, but I would do the hotfix for the components I need right now
b) yes, I can help you with the declarations, if you want

**Comment by prateekbh** [comment]
@cromefire Thanks so much for understanding.
Lets revert your last commit. and open a new PR to shift everything to angular's `fesm` style with the help of rollup and `babel-preset-env`. Makes sense?

**Comment by cromefire** [comment]
Okay, thanks to regex search I fixed all .jsx (+.d.ts) files. This should be okay until you change the build system

File: compileComponents.js [hunk 1]
```diff
@@ -4,11 +4,30 @@ const bundleMapping = require('./componentsList');
 const path = require('path');
 const fs = require('fs');
 
+const component_array = [];
```
Comment (prateekbh) [inline compileComponents.js]: lets revert this file and i'll merge this in and release it

File: Card/Card.jsx [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import Button from '../Button';
-import Icon from '../Icon';
+import Button from '../Button/Button';
```
Comment (prateekbh) [inline Card/Card.jsx]: dont make this change in this PR

File: compileComponents.js (hunk 1 above)
Comment (cromefire) [reply compileComponents.js]: Will do that in a few minutes

File: Card/Card.jsx (hunk 2 above)
Comment (cromefire) [reply Card/Card.jsx]: I hadn't read your response will revert that

Comment (cromefire) [reply Card/Card.jsx]: I'm currently on slow internet

File: Chips/Chips.jsx [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import {h} from 'preact';
-import Icon from '../Icon';
+import Icon from '../Icon/Icon';
```
Comment (prateekbh) [inline Chips/Chips.jsx]: Revert this, lets fix this in a separate PR

File: Dialog/Dialog.jsx [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
 import {MDCDialog} from '@material/dialog/';
-import Button from '../Button';
+import Button from '../Button/Button';
```
Comment (prateekbh) [inline Dialog/Dialog.jsx]: Revert

File: Drawer/Drawer.jsx [hunk 5]
```diff
@@ -2,7 +2,7 @@ import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
 import {MDCTemporaryDrawer} from '@material/drawer/temporary';
 import {MDCPersistentDrawer} from '@material/drawer/persistent';
-import List from '../List';
+import List from '../List/List';
```
Comment (prateekbh) [inline Drawer/Drawer.jsx]: Revert

File: Fab/Fab.jsx [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import Icon from '../Icon/';
+import Icon from '../Icon/Icon';
```
Comment (prateekbh) [inline Fab/Fab.jsx]: REvert

File: GridList/GridList.jsx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
-import Icon from '../Icon';
+import Icon from '../Icon/Icon';
```
Comment (prateekbh) [inline GridList/GridList.jsx]: Revert

File: Menu/Menu.jsx [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
 import {MDCMenu} from '@material/menu';
-import List from '../List';
+import List from '../List/List';
```
Comment (prateekbh) [inline Menu/Menu.jsx]: REvert

File: Select/Select.jsx [hunk 9]
```diff
@@ -1,8 +1,8 @@
 import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
 import {MDCSelect} from '@material/select/';
-import List from '../List';
-import Menu from '../Menu';
+import List from '../List/List';
```
Comment (prateekbh) [inline Select/Select.jsx]: Revert this file

File: TextField/TextField.jsx [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import {Component, h} from 'preact';
 
-import Icon from '../Icon';
+import Icon from '../Icon/Icon';
```
Comment (prateekbh) [inline TextField/TextField.jsx]: Revert file

File: List/List.jsx [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import {h} from 'preact';
 import MaterialComponent from '../MaterialComponent';
-import Icon from '../Icon';
+import Icon from '../Icon/Icon';
```
Comment (prateekbh) [inline List/List.jsx]: Revert this file

**Comment by cromefire** [comment]
If someone of you has time, edits from maintainers are allowed so you can just revert the last 3 commits, else I'll do that in about an hour

**Comment by prateekbh** [comment]
@cromefire lets make another PR which generates `index.js` and `index.esm.js` for ever component in its own folder.

Also we'll discuss how to do TS definition for the same in that PR.

**Comment by cromefire** [comment]
Okay let's do that, closing this one.

**Comment by prateekbh** [comment]
sure go for it...
10 foot overview: 
a.) Use rollup to create index.js and index.esm.js
b.) figure out which TS files would be used when we do this change.
c.)  tell users to not use .jsx and use .esm.js or .js instead.

Sounds fine?

**Comment by cromefire** [comment]
a) I would think we should evaluate which build system is the best for our use case
b)
     

- You have to decide that and it's just a suggestion, but we could change the files to typescript (it's really not that hard and for the beginning you could start with just renaming files to .tsx and the rest is just improving. Prettier and eslint should be fine with that, maybe adding(/replacing eslint) tslint would be okay). This could solve a) and b).
- Else it shouldn't be that hard, I can do that
c) fine

**Comment by prateekbh** [comment]
>You have to decide that and it's just a suggestion, but we could change the files to typescript (it's really not that hard and for the beginning you could start with just renaming files to .tsx 

What happens to people trying to use TS and want to use a both transpiled version `es5` and `esm`

**Comment by prateekbh** [comment]
Also, rollup has the best minification metric so far

**Comment by cromefire** [comment]
> What happens to people trying to use TS and want to use a both transpiled version es5 and esm

For those who want to use the transpiled file, they can just import the js file, typescript can generate definition files for them. And about `esm`, I didn't got the difference between `esm` and original

> Also, rollup has the best minification metric so far

But do we need a bundler?
I think we just need to transpile to es5 and `cjs`, right?

**Comment by prateekbh** [comment]
>For those who want to use the transpiled file, they can just import the js file, typescript can generate definition files for them. And about esm, I didn't got the difference between esm and original

`.esm.js` files has the code that you'd like to send in `<script type=module/ES2015` and `.js` files has  will have es5 code.


>I think we just need to transpile to es5 and cjs, right?
Makes sense, but lets just transpile down the files to ESM and ES5 and convert the imports from `../X/X` to `../X`

**Comment by cromefire** [comment]
> .esm.js files has the code that you'd like to send in <script type=module/ES2015

That would be what typescript emits

> Makes sense, but lets just transpile down the files to ESM and ES5 and convert the imports from ../X/X to ../X

Wouldn't it make sense to do this with directories?
`ts/` => sources
`./` => for `es5`
`esm/` => for esm

we would need something to transpile the esm files, but generating the esm files could be done directly by typescript (if you agree on using typescript of course).

**Comment by cromefire** [comment]
And it would be nice if you could create a new branch and the PR, I'm still cloning the repo.

**Comment by prateekbh** [comment]
@cromefire i sent you a collaborator invite, specifically so that you can do this

**Comment by prateekbh** [comment]
>Wouldn't it make sense to do this with directories?
ts/ => sources
./ => for es5
esm/ => for esm

Sounds perfect

**Comment by cromefire** [comment]
@prateekbh you did, but my internet will be very slow this week, so I'm waiting on the download

**Comment by prateekbh** [comment]
here, use this branch `build-esm-output`

**Pull request closed without merging**